File: archive/v1-dashboard/tools/fsociety.py
#!/usr/bin/env python3
"""Generate a glitching fsociety-style mask animation as fsociety.anim.

Pure standard library - no Pillow needed. The mask is drawn from geometric
primitives at 160x80 (the T-Dongle-C5 panel in landscape), then a scanline +
slice-glitch + channel-split pass is applied per frame.

    python3 tools/fsociety.py            # -> fsociety.anim
    python3 tools/fsociety.py out.anim
"""
import os
import random
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from animfmt import write_anim, pack_frame, load_font, draw_text  # noqa: E402

W, H = 160, 80
FRAMES = 48
DELAY_MS = 65

# Mask geometry, in pixels relative to the face centre.
CX, CY = 80.0, 30.0
A, B_TOP, B_BOT = 19.5, 26.0, 29.0

BG      = (0, 0, 0)
FACE    = (232, 220, 198)
DARK    = (16, 13, 11)
SHADOW  = (186, 173, 152)
OUTLINE = (60, 52, 44)
LIME    = (0, 255, 65)

random.seed(0xF50C)


def ellipse(dx, dy, ex, ey, rx, ry):
    if rx <= 0 or ry <= 0:
        return False
    u = (dx - ex) / rx
    v = (dy - ey) / ry
    return u * u + v * v <= 1.0


def in_face(dx, dy):
    b = B_TOP if dy < 0 else B_BOT
    t = dy / b
    if abs(t) > 1.0:
        return False
    a_eff = A * (1.0 - 0.26 * max(0.0, t) ** 1.9)   # taper toward the chin
    u = dx / a_eff
    return u * u + t * t <= 1.0


def in_brow(dx, dy):
    # A band between two concentric ellipses, split into a left and right arch.
    # Kept well above the eyes so the two never merge into one dark mass.
    inner = ellipse(dx, dy, 0, -3.0, 17.0, 12.6)
    outer = ellipse(dx, dy, 0, -3.0, 17.0, 15.8)
    return outer and not inner and dy < -12.6 and 3.0 <= abs(dx) <= 15.0


def in_eye(dx, dy):
    return ellipse(dx, dy, -8.6, -5.0, 5.6, 6.6) or ellipse(dx, dy, 8.6, -5.0, 5.6, 6.6)


def in_mustache(dx, dy):
    # Two wings joined by a thin bridge, so a dip stays visible under the nose,
    # plus the upturned curls at the outer ends.
    return (ellipse(dx, dy, -8.5, 9.5, 8.2, 3.9)
            or ellipse(dx, dy, 8.5, 9.5, 8.2, 3.9)
            or ellipse(dx, dy, 0.0, 8.4, 3.8, 2.0)
            or ellipse(dx, dy, -15.8, 5.8, 3.2, 3.6)
            or ellipse(dx, dy, 15.8, 5.8, 3.2, 3.6))


def in_goatee(dx, dy):
    # Sits clear of the mustache with face showing between the two.
    return ellipse(dx, dy, 0, 20.5, 4.0, 4.6) or ellipse(dx, dy, 0, 24.5, 2.3, 4.2)


def build_base():
    px = [BG] * (W * H)
    for y in range(H):
        dy = y - CY
        for x in range(W):
            dx = x - CX
            if not in_face(dx, dy):
                continue
            if in_eye(dx, dy) or in_brow(dx, dy) or in_mustache(dx, dy) \
                    or in_goatee(dx, dy):
                c = DARK
            else:
                c = FACE
            px[y * W + x] = c

    # 1px outline so the pale face reads against the black background.
    edged = list(px)
    for y in range(H):
        for x in range(W):
            if px[y * W + x] != BG:
                continue
            for ny, nx in ((y - 1, x), (y + 1, x), (y, x - 1), (y, x + 1)):
                if 0 <= ny < H and 0 <= nx < W and px[ny * W + nx] == FACE:
                    edged[y * W + x] = OUTLINE
                    break
    return edged


def scanline(px):
    for y in range(1, H, 2):
        base = y * W
        for x in range(W):
            r, g, b = px[base + x]
            px[base + x] = (r * 78 // 100, g * 78 // 100, b * 78 // 100)


def slice_glitch(px, strength):
    for _ in range(strength):
        y0 = random.randrange(0, H - 2)
        hh = random.randint(2, 9)
        shift = random.randint(-14, 14)
        if shift == 0:
            continue
        for y in range(y0, min(H, y0 + hh)):
            row = px[y * W:(y + 1) * W]
            px[y * W:(y + 1) * W] = row[-shift:] + row[:-shift]


def channel_split(px, offset):
    src = list(px)
    for y in range(H):
        for x in range(W):
            r = src[y * W + max(0, x - offset)][0]
            g = src[y * W + x][1]
            b = src[y * W + min(W - 1, x + offset)][2]
            px[y * W + x] = (r, g, b)


def noise(px, prob):
    n = int(W * H * prob)
    for _ in range(n):
        i = random.randrange(W * H)
        px[i] = LIME if random.random() < 0.35 else (200, 200, 200)


def main():
    out = sys.argv[1] if len(sys.argv) > 1 else "fsociety.anim"
    font = load_font(os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                  "..", "main", "font5x8.h"))
    base = build_base()

    frames = []
    for f in range(FRAMES):
        px = list(base)

        # --- wordmark types itself in during the first second ---------------
        letters = "fsociety"
        shown = len(letters) if f >= 16 else max(0, (f - 2) // 2)
        if shown:
            draw_text(px, W, H, 32, 63, letters[:shown], LIME, font, scale=2)
        if 6 <= f < 16 and f % 2 == 0 and shown < len(letters):
            draw_text(px, W, H, 32 + shown * 12, 63, "_", LIME, font, scale=2)

        # --- reveal, hold, then break apart ---------------------------------
        if f < 8:
            cutoff = int(H * (f + 1) / 8.0)
            for y in range(cutoff, H):
                for x in range(W):
                    px[y * W + x] = BG
            noise(px, 0.10)
            slice_glitch(px, 3)
        elif f >= FRAMES - 4:
            slice_glitch(px, 5)
            channel_split(px, 3)
            noise(px, 0.05)
        elif f % 11 in (0, 1):
            slice_glitch(px, 2)
            channel_split(px, 2)
        elif f % 17 == 5:
            noise(px, 0.012)

        scanline(px)
        frames.append((DELAY_MS, pack_frame(px, W, H)))

    write_anim(out, W, H, frames, DELAY_MS)


if __name__ == "__main__":
    main()

File: archive/v1-dashboard/main/fsociety.c
#include "fsociety.h"
#include "gfx.h"
#include "board.h"
#include "esp_heap_caps.h"
#include "esp_log.h"
#include <string.h>
#include <stdlib.h>

static const char *TAG = "fsociety";

// Face geometry, in pixels relative to the centre of the mask. Mirrors
// tools/fsociety.py so the on-device render and the .anim export agree.
#define CX      80.0f
#define CY      30.0f
#define A       19.5f
#define B_TOP   26.0f
#define B_BOT   29.0f

#define FRAME_COUNT 48
#define FRAME_MS    65

#define C_FACE    RGB(232, 220, 198)
#define C_DARK    RGB(16,  13,  11)
#define C_OUTLINE RGB(60,  52,  44)

static uint16_t *s_base = NULL;      // the clean mask, rendered once
static uint16_t  s_frame = 0;
static uint8_t   dim5[32], dim6[64]; // scanline dimming LUTs

// Small xorshift so the glitch pattern is cheap and repeatable across boots.
static uint32_t s_rng = 0xF50C1E7Fu;
static inline uint32_t rnd(void)
{
    s_rng ^= s_rng << 13;
    s_rng ^= s_rng >> 17;
    s_rng ^= s_rng << 5;
    return s_rng;
}

static inline uint16_t sw(uint16_t v) { return (uint16_t)((v >> 8) | (v << 8)); }

static inline bool ellipse(float dx, float dy, float ex, float ey, float rx, float ry)
{
    float u = (dx - ex) / rx;
    float v = (dy - ey) / ry;
    return u * u + v * v <= 1.0f;
}

static bool in_face(float dx, float dy)
{
    float b = (dy < 0) ? B_TOP : B_BOT;
    float t = dy / b;
    if (t < -1.0f || t > 1.0f) return false;
    float taper = t > 0 ? t : 0.0f;
    float a_eff = A * (1.0f - 0.26f * taper * taper);   // jaw narrows toward the chin
    float u = dx / a_eff;
    return u * u + t * t <= 1.0f;
}

static bool in_brow(float dx, float dy)
{
    // Band between two concentric ellipses, split into a left and right arch and
    // kept clear of the eyes so the two never merge into one dark mass.
    bool inner = ellipse(dx, dy, 0, -3.0f, 17.0f, 12.6f);
    bool outer = ellipse(dx, dy, 0, -3.0f, 17.0f, 15.8f);
    float adx = dx < 0 ? -dx : dx;
    return outer && !inner && dy < -12.6f && adx >= 3.0f && adx <= 15.0f;
}

static bool in_eye(float dx, float dy)
{
    return ellipse(dx, dy, -8.6f, -5.0f, 5.6f, 6.6f)
        || ellipse(dx, dy,  8.6f, -5.0f, 5.6f, 6.6f);
}

static bool in_mustache(float dx, float dy)
{
    // Two wings joined by a thin bridge so a dip shows under the nose, plus the
    // upturned curls at the outer ends.
    return ellipse(dx, dy, -8.5f, 9.5f, 8.2f, 3.9f)
        || ellipse(dx, dy,  8.5f, 9.5f, 8.2f, 3.9f)
        || ellipse(dx, dy,  0.0f, 8.4f, 3.8f, 2.0f)
        || ellipse(dx, dy, -15.8f, 5.8f, 3.2f, 3.6f)
        || ellipse(dx, dy,  15.8f, 5.8f, 3.2f, 3.6f);
}

static bool in_goatee(float dx, float dy)
{
    return ellipse(dx, dy, 0, 20.5f, 4.0f, 4.6f)
        || ellipse(dx, dy, 0, 24.5f, 2.3f, 4.2f);
}

esp_err_t fsociety_init(void)
{
    if (s_base) return ESP_OK;

    s_base = heap_caps_malloc(LCD_W * LCD_H * 2, MALLOC_CAP_SPIRAM | MALLOC_CAP_8BIT);
    if (!s_base) s_base = malloc(LCD_W * LCD_H * 2);
    if (!s_base) return ESP_ERR_NO_MEM;

    for (int i = 0; i < 32; i++) dim5[i] = (uint8_t)((i * 78) / 100);
    for (int i = 0; i < 64; i++) dim6[i] = (uint8_t)((i * 78) / 100);

    // Soft float on the C5 is slow, but this runs exactly once.
    for (int y = 0; y < LCD_H; y++) {
        float dy = (float)y - CY;
        for (int x = 0; x < LCD_W; x++) {
            float dx = (float)x - CX;
            uint16_t c = C_BLACK;
            if (in_face(dx, dy)) {
                c = (in_eye(dx, dy) || in_brow(dx, dy) || in_mustache(dx, dy)
                     || in_goatee(dx, dy)) ? C_DARK : C_FACE;
            }
            s_base[y * LCD_W + x] = c;
        }
    }

    // 1px outline so the pale face reads against black.
    for (int y = 0; y < LCD_H; y++) {
        for (int x = 0; x < LCD_W; x++) {
            if (s_base[y * LCD_W + x] != C_BLACK) continue;
            bool touches = false;
            if (y > 0          && s_base[(y - 1) * LCD_W + x] == C_FACE) touches = true;
            if (y < LCD_H - 1  && s_base[(y + 1) * LCD_W + x] == C_FACE) touches = true;
            if (x > 0          && s_base[y * LCD_W + x - 1]   == C_FACE) touches = true;
            if (x < LCD_W - 1  && s_base[y * LCD_W + x + 1]   == C_FACE) touches = true;
            if (touches) s_base[y * LCD_W + x] = C_OUTLINE;
        }
    }

    ESP_LOGI(TAG, "mask rasterised");
    return ESP_OK;
}

// ------------------------------------------------------------- effects ----
static void slice_glitch(int count)
{
    static uint16_t row[LCD_W];
    for (int i = 0; i < count; i++) {
        int y0 = rnd() % (LCD_H - 2);
        int hh = 2 + rnd() % 8;
        int shift = (int)(rnd() % 29) - 14;
        if (!shift) continue;
        for (int y = y0; y < y0 + hh && y < LCD_H; y++) {
            uint16_t *src = &g_fb[y * LCD_W];
            for (int x = 0; x < LCD_W; x++) {
                int sx = x - shift;
                while (sx < 0) sx += LCD_W;
                while (sx >= LCD_W) sx -= LCD_W;
                row[x] = src[sx];
            }
            memcpy(src, row, sizeof(row));
        }
    }
}

static void channel_split(int off)
{
    static uint16_t row[LCD_W];
    for (int y = 0; y < LCD_H; y++) {
        uint16_t *line = &g_fb[y * LCD_W];
        memcpy(row, line, sizeof(row));
        for (int x = 0; x < LCD_W; x++) {
            int xr = x - off; if (xr < 0) xr = 0;
            int xb = x + off; if (xb >= LCD_W) xb = LCD_W - 1;
            uint16_t r = sw(row[xr]) & 0xF800;
            uint16_t g = sw(row[x])  & 0x07E0;
            uint16_t b = sw(row[xb]) & 0x001F;
            line[x] = sw(r | g | b);
        }
    }
}

static void noise(int per_thousand)
{
    int n = (LCD_W * LCD_H * per_thousand) / 1000;
    for (int i = 0; i < n; i++) {
        uint32_t r = rnd();
        g_fb[r % (LCD_W * LCD_H)] = ((r >> 16) & 0xFF) < 90 ? C_LIME : RGB(200, 200, 200);
    }
}

static void scanlines(void)
{
    for (int y = 1; y < LCD_H; y += 2) {
        uint16_t *line = &g_fb[y * LCD_W];
        for (int x = 0; x < LCD_W; x++) {
            uint16_t n = sw(line[x]);
            if (!n) continue;
            uint16_t v = ((uint16_t)dim5[(n >> 11) & 0x1F] << 11)
                       | ((uint16_t)dim6[(n >> 5)  & 0x3F] << 5)
                       |  (uint16_t)dim5[n & 0x1F];
            line[x] = sw(v);
        }
    }
}

// -------------------------------------------------------------- render ----
uint16_t fsociety_draw(void)
{
    if (!s_base && fsociety_init() != ESP_OK) {
        gfx_clear(C_BLACK);
        gfx_text(20, 36, "mask alloc failed", C_RED, 1);
        return 1000;
    }

    const uint16_t f = s_frame;
    memcpy(g_fb, s_base, LCD_W * LCD_H * 2);

    // The wordmark types itself in, with a blinking caret.
    static const char WORD[] = "fsociety";
    int shown = (f >= 16) ? 8 : (int)((f > 2 ? f - 2 : 0) / 2);
    if (shown > 8) shown = 8;
    if (shown > 0) {
        char buf[9];
        memcpy(buf, WORD, shown);
        buf[shown] = '\0';
        gfx_text(32, 63, buf, C_LIME, 2);
    }
    if (f >= 6 && f < 16 && (f & 1) == 0 && shown < 8) {
        gfx_text(32 + shown * 12, 63, "_", C_LIME, 2);
    }

    if (f < 8) {                            // resolve out of static
        int cutoff = (LCD_H * (f + 1)) / 8;
        for (int y = cutoff; y < LCD_H; y++)
            memset(&g_fb[y * LCD_W], 0, LCD_W * 2);
        noise(55);
        slice_glitch(3);
    } else if (f >= FRAME_COUNT - 4) {      // break apart before looping
        slice_glitch(5);
        channel_split(3);
        noise(28);
    } else if (f % 11 <= 1) {
        slice_glitch(2);
        channel_split(2);
    } else if (f % 17 == 5) {
        noise(8);
    }

    scanlines();

    s_frame = (uint16_t)((f + 1) % FRAME_COUNT);
    return FRAME_MS;
}
